forum/src/main/java/unipd/se18/ocrcamera/forum/ForumLogin.java

```diff
@@ -0,0 +1,174 @@
+    }
+
+    @Override
+    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
+        return inflater.inflate(R.layout.fragment_forum_login, container, false);
+    }
+
+    @Override
+    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
+
+        super.onViewCreated(view, savedInstanceState);
+
+        //UI object initialization
+        usernameEditText = view.findViewById(R.id.usernameEditText);
+        pwdEditText = view.findViewById(R.id.pwdEditText);
+        loginButton = view.findViewById(R.id.loginButton);
+
+        //Definition of view model listener
+        viewModel.setForumLoginListener(new Login_VM.ForumLoginListener() {
+
+            /**
+             * The method of the listener is triggered when the database response
+             * to the login request is positive and the view model requests to load
+             * the ShowPost fragment, so that the user can access the forum contents
+             *
+             * @param username The username of the user that successfully logged in
+             * @author Alberto Valente (g2), Taulant Bullaku (g2)
+             */
+            @Override
+            public void onLoginSuccess(String username) {
+
+                if(username != null) {
+                    //creates the username bundle
+                    Bundle bundle = new Bundle();
+                    bundle.putString(KEY_USERNAME, username);
+
+                    //creates an instance of the fragment to be launched
+                    ShowPosts showPostsFragment = new ShowPosts();
+                    //passes the bundle to the fragment as an argument
+                    showPostsFragment.setArguments(bundle);
+
+                    //performs the fragment transaction
+                    getActivity()
+                            .getSupportFragmentManager()
+                            .beginTransaction()
+                            .replace(
+                                    R.id.fragmentContainer,
+                                    showPostsFragment
+                            )
+                            .commit();
+                }
+                else {
+                    Log.d(LOG_TAG, LOG_NULL_USERNAME);
+                }
+            }
+
+            /**
+             * The method of the listener is triggered when the database response to
+             * the login request is negative and the view model requests to show the
+             * user an error message, that is performed through the use of a toast
+             *
+             * @param message The error message about what was wrong with the login request
+             * @author Alberto Valente (g2), Taulant Bullaku (g2)
+             */
+            @Override
+            public void onLoginFailure(String message) {
+
+                if (message != null) {
+                    Toast.makeText(getContext(), message, Toast.LENGTH_LONG).show();
```

Comment:
Here I can't really understand where the content of the String "message" comes from. Is it something in the "Login_VM" class? I didn't seem to find anything there though.